Add LIP "Add weights to Lisk-BFT consensus protocol" (#111)
LiskHQ/lips: proposals/lip-add_weights_to_lisk-bft_consensus_protocol.md

## Copyright

This LIP is licensed under the [Creative Commons Zero 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/).

## Motivation

The Lisk-BFT consensus protocol introduced in [LIP 0014][lisk-bft-lip] was specified for the Lisk mainchain and sidechains that use Delegated Proof-of-Stake with fixed configuration parameters that only change in rare cases as part of a hard fork.
In these blockchains, there is a fixed number of active delegates, i.e., those delegates participating in finalizing blocks, and all of them have equal weight.
One exception to that is the bootstrap period introduced as part of the new genesis block format in [LIP 0034][lip-34].
During the bootstrap period none of the bootstrap delegates contributes to finalizing blocks, so that blocks can only be finalized after the bootstrap period is over.

For more flexibility and to also support other validator selection mechanisms, we want to allow for different BFT weights of the active validators, i.e., different weights attributed to the prevotes and precommits cast by a validator.

> These weights and the threshold for finalizing a block should also not be constant, but should be allowed to change over time. This is, in particular, required in conjunction with the Proof-of-Authority validator selection mechanism introduced in the [PoA][module] where validators can be added and removed or their weights changed with a simple transaction and without a hard fork.

Reviewer: This link to the PoA module seems broken.

Author: I fixed the link.

A transformation from the BFT weights and thresholds in this LIP to the BFT weights and thresholds used in the paper is done by dividing all BFT weights and thresholds at height `h` by the sum of BFT weights at height `h`.

The Lisk-BFT protocol introduced in Definition 4.1 of the [Lisk-BFT paper][lisk-bft-paper] allows to choose two protocol parameters, `τ ∈ (1/3, 1]` and a natural number `ρ`.
The decision threshold `τ` in the paper corresponds to the precommit threshold used in this LIP up to rescaling and the subtle difference that the weight of precommits in the paper has to be strictly more than `τ` but in these specifications it only has to be greater or equal than `precommitThreshold(h)` (as we restrict the range of `precommitThreshold(h)` such that `precommitThreshold(h) / aggregateBFTWeight(h) > 1/3`).
Similarly, we choose `prevoteThreshold(h)` such that `prevoteThreshold(h) / aggregateBFTWeight(h) > 2/3`.

The parameter `ρ`, which determines how many blocks a block can imply prevotes for, is given by `3*LSK_BFT_BATCH_SIZE` in the specifications in this LIP, as also in LIP 0014.
Here `LSK_BFT_BATCH_SIZE` is a constant of the  [BFT module LIP][bft-module-lip].

> As the round length in a PoA chain can vary, we require that the maximum possible round length in a PoA chain, given by the constant `MAX_NUM_VALIDATORS` defined in the [Proof-of-Authority LIP][poa-lip], is always at most the constant `LSK_BFT_BATCH_SIZE`.

Reviewer 2: ```suggestion
As the round length in a PoA chain can vary, we require that the maximum possible round length in a PoA chain, given by the constant `MAX_NUM_VALIDATORS` defined in [LIP 0047][poa-lip], is always at most the constant `LSK_BFT_BATCH_SIZE`.
```

Author: changed to "defined in the PoA module" instead

The reason is that it is not uniquely determined by the tip of the chain, but depends on the history of executed and reverted blocks as explained in the rationale at the beginning.
The value of `maxHeightFinalized` can be easily updated as follows:
```
maxHeightFinalized = max(bftVotes.maxHeightPrecommitted, maxHeightFinalized).
```
In particular, it only needs to be updated if `bftVotes.maxHeightPrecommitted` increases.

## Backwards Compatibility

This LIP defines a part of the state transition logic of the BFT module introduced in the [BFT module LIP][bft-module-lip].
The LIP therefore requires a hard fork as introducing the BFT module implies a hardfork.

[bft-module-lip]: https://research.lisk.com/t/introduce-bft-module/321
[lip-34]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0034.md
[lisk-bft-lip]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md
[lisk-bft-lip-comp]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#computing-prevotes-and-precommits
[lisk-bft-paper]: https://arxiv.org/abs/1903.11434

> [poa-lip]: https://research.lisk.com/t/proof-of-authority-validator-selection-mechanism/288

Reviewer 2: ```suggestion
[poa-lip]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0047.md
```

Author: Fixed.

> [validator-module-lip]: https://research.lisk.com/t/introduce-validators-module/317

Reviewer 2: ```suggestion
[validator-module-lip]: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0044.md
```

Author: Fixed.
